File: apps/api/src/modules/auth/application/token.service.ts
import { createHash, randomUUID } from 'node:crypto';

import { Injectable, Logger } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';

import { AppConfigService } from '../../../common/config/app-config.service';
import { ERROR_CODES } from '../../../common/errors/error-codes';
import { UnauthorizedError } from '../../../common/errors/app.exception';
import { PrismaService } from '../../../common/prisma/prisma.service';
import type {
  AccessTokenPayload,
  AuthUser,
  RefreshTokenPayload,
} from '../domain/auth-user';

export interface IssuedRefreshToken {
  token: string;
  expiresAt: Date;
}

@Injectable()
export class TokenService {
  private readonly logger = new Logger(TokenService.name);

  constructor(
    private readonly jwt: JwtService,
    private readonly config: AppConfigService,
    private readonly prisma: PrismaService,
  ) {}

  /**
   * Access tokens are short lived and never stored: verifying one is a pure
   * signature check, which is what keeps authenticated reads cheap.
   */
  async signAccessToken(user: Pick<AuthUser, 'id' | 'username' | 'role'>): Promise<string> {
    const payload: AccessTokenPayload = {
      sub: user.id,
      username: user.username,
      role: user.role,
      type: 'access',
    };

    return this.jwt.signAsync(payload, {
      secret: this.config.jwt.accessSecret,
      expiresIn: this.config.jwt.accessTtlSeconds,
    });
  }

  async verifyAccessToken(token: string): Promise<AccessTokenPayload> {
    try {
      const payload = await this.jwt.verifyAsync<AccessTokenPayload>(token, {
        secret: this.config.jwt.accessSecret,
      });

      if (payload.type !== 'access') {
        throw new UnauthorizedError(ERROR_CODES.UNAUTHORIZED, 'Wrong token type');
      }
      return payload;
    } catch (error) {
      if (error instanceof UnauthorizedError) throw error;

      const expired = error instanceof Error && error.name === 'TokenExpiredError';
      throw new UnauthorizedError(
        expired ? ERROR_CODES.ACCESS_TOKEN_EXPIRED : ERROR_CODES.UNAUTHORIZED,
        expired ? 'Access token expired' : 'Invalid access token',
      );
    }
  }

  /**
   * Refresh tokens are signed JWTs whose SHA-256 hash is stored in the
   * database. The signature makes them cheap to reject, and the stored hash
   * makes them revocable; storing only the hash means a database leak cannot be
   * replayed as a live session.
   */
  async issueRefreshToken(userId: string, userAgent?: string): Promise<IssuedRefreshToken> {
    const jti = randomUUID();
    const ttlSeconds = this.config.jwt.refreshTtlSeconds;
    const payload: RefreshTokenPayload = { sub: userId, jti, type: 'refresh' };

    const token = await this.jwt.signAsync(payload, {
      secret: this.config.jwt.refreshSecret,
      expiresIn: ttlSeconds,
    });

    const expiresAt = new Date(Date.now() + ttlSeconds * 1000);

    await this.prisma.refreshToken.create({
      data: {
        id: jti,
        userId,
        tokenHash: this.hashToken(token),
        expiresAt,
        userAgent: userAgent?.slice(0, 255) ?? null,
      },
    });

    return { token, expiresAt };
  }

  /**
   * Validates a refresh token and replaces it with a new one (rotation).
   *
   * A token that verifies but is missing or already revoked in the database has
   * been replayed, which means it probably leaked. In that case every session
   * for the user is revoked instead of just refusing this one request.
   */
  async rotateRefreshToken(
    rawToken: string,
    userAgent?: string,
  ): Promise<{ userId: string; refreshToken: IssuedRefreshToken }> {
    const payload = await this.verifyRefreshToken(rawToken);
    const tokenHash = this.hashToken(rawToken);

    const stored = await this.prisma.refreshToken.findUnique({
      where: { tokenHash },
      select: { id: true, userId: true, revokedAt: true, expiresAt: true },
    });

    if (!stored) {
      await this.revokeAllForUser(payload.sub);
      this.logger.warn(
        `Refresh token for user ${payload.sub} was not found; all sessions revoked.`,
      );
      throw new UnauthorizedError(ERROR_CODES.INVALID_REFRESH_TOKEN, 'Refresh token is not valid');
    }

    if (stored.revokedAt !== null) {
      await this.revokeAllForUser(stored.userId);
      this.logger.warn(
        `Reuse of a revoked refresh token by user ${stored.userId}; all sessions revoked.`,
      );
      throw new UnauthorizedError(
        ERROR_CODES.INVALID_REFRESH_TOKEN,
        'Refresh token has already been used',
      );
    }

    if (stored.expiresAt.getTime() <= Date.now()) {
      throw new UnauthorizedError(ERROR_CODES.INVALID_REFRESH_TOKEN, 'Refresh token has expired');
    }

    const refreshToken = await this.prisma.$transaction(async (tx) => {
      await tx.refreshToken.update({
        where: { id: stored.id },
        data: { revokedAt: new Date() },
      });

      const jti = randomUUID();
      const ttlSeconds = this.config.jwt.refreshTtlSeconds;
      const token = await this.jwt.signAsync(
        { sub: stored.userId, jti, type: 'refresh' } satisfies RefreshTokenPayload,
        { secret: this.config.jwt.refreshSecret, expiresIn: ttlSeconds },
      );
      const expiresAt = new Date(Date.now() + ttlSeconds * 1000);

      await tx.refreshToken.create({
        data: {
          id: jti,
          userId: stored.userId,
          tokenHash: this.hashToken(token),
          expiresAt,
          userAgent: userAgent?.slice(0, 255) ?? null,
        },
      });

      return { token, expiresAt };
    });

    return { userId: stored.userId, refreshToken };
  }

  /** Idempotent: logging out twice is not an error. */
  async revokeRefreshToken(rawToken: string): Promise<void> {
    await this.prisma.refreshToken.updateMany({
      where: { tokenHash: this.hashToken(rawToken), revokedAt: null },
      data: { revokedAt: new Date() },
    });
  }

  async revokeAllForUser(userId: string): Promise<void> {
    await this.prisma.refreshToken.updateMany({
      where: { userId, revokedAt: null },
      data: { revokedAt: new Date() },
    });
  }

  /** Housekeeping for the scheduled maintenance job. */
  async deleteExpiredTokens(): Promise<number> {
    const result = await this.prisma.refreshToken.deleteMany({
      where: { expiresAt: { lt: new Date() } },
    });
    return result.count;
  }

  private async verifyRefreshToken(token: string): Promise<RefreshTokenPayload> {
    try {
      const payload = await this.jwt.verifyAsync<RefreshTokenPayload>(token, {
        secret: this.config.jwt.refreshSecret,
      });
      if (payload.type !== 'refresh') {
        throw new UnauthorizedError(ERROR_CODES.INVALID_REFRESH_TOKEN, 'Wrong token type');
      }
      return payload;
    } catch (error) {
      if (error instanceof UnauthorizedError) throw error;
      throw new UnauthorizedError(ERROR_CODES.INVALID_REFRESH_TOKEN, 'Refresh token is not valid');
    }
  }

  private hashToken(token: string): string {
    return createHash('sha256').update(token).digest('hex');
  }
}

File: apps/api/src/GameScore.Infrastructure/Auth/TokenService.cs
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;
using GameScore.Application.Auth;
using GameScore.Application.Configuration;
using GameScore.Application.Errors;
using GameScore.Infrastructure.Crypto;
using GameScore.Infrastructure.Persistence;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using Microsoft.IdentityModel.Tokens;

namespace GameScore.Infrastructure.Auth;

public sealed class TokenService
{
    private readonly AppConfig _config;
    private readonly GameScoreDbContext _db;
    private readonly ILogger<TokenService> _logger;

    public TokenService(IOptions<AppConfig> config, GameScoreDbContext db, ILogger<TokenService> logger)
    {
        _config = config.Value;
        _db = db;
        _logger = logger;
    }

    public string SignAccessToken(Guid userId, string username, string role, int reputationScore)
    {
        var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(_config.Jwt.AccessSecret));
        var credentials = new SigningCredentials(key, SecurityAlgorithms.HmacSha256);
        var claims = new[]
        {
            new Claim(JwtRegisteredClaimNames.Sub, userId.ToString()),
            new Claim("username", username),
            new Claim("role", role),
            new Claim("reputationScore", reputationScore.ToString()),
            new Claim("type", "access"),
        };

        var token = new JwtSecurityToken(
            claims: claims,
            expires: DateTime.UtcNow.AddSeconds(_config.Jwt.AccessTtlSeconds),
            signingCredentials: credentials);

        return new JwtSecurityTokenHandler().WriteToken(token);
    }

    public AccessTokenPayload VerifyAccessToken(string token)
    {
        try
        {
            var handler = new JwtSecurityTokenHandler();
            var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(_config.Jwt.AccessSecret));
            var principal = handler.ValidateToken(
                token,
                new TokenValidationParameters
                {
                    ValidateIssuer = false,
                    ValidateAudience = false,
                    ValidateIssuerSigningKey = true,
                    IssuerSigningKey = key,
                    ClockSkew = TimeSpan.FromSeconds(30),
                },
                out _);

            var type = principal.FindFirst("type")?.Value;
            if (type != "access")
            {
                throw new UnauthorizedException(ErrorCodes.Unauthorized, "Wrong token type");
            }

            var sub = principal.FindFirst(JwtRegisteredClaimNames.Sub)?.Value
                ?? principal.FindFirst(ClaimTypes.NameIdentifier)?.Value;
            if (sub is null)
            {
                throw new UnauthorizedException(ErrorCodes.Unauthorized, "Invalid access token");
            }

            return new AccessTokenPayload(
                sub,
                principal.FindFirst("username")?.Value ?? "",
                principal.FindFirst("role")?.Value ?? "USER",
                "access");
        }
        catch (SecurityTokenExpiredException)
        {
            throw new UnauthorizedException(ErrorCodes.AccessTokenExpired, "Access token expired");
        }
        catch (UnauthorizedException)
        {
            throw;
        }
        catch
        {
            throw new UnauthorizedException(ErrorCodes.Unauthorized, "Invalid access token");
        }
    }

    public async Task<IssuedRefreshToken> IssueRefreshTokenAsync(
        Guid userId,
        string? userAgent,
        CancellationToken cancellationToken = default)
    {
        var jti = Guid.NewGuid();
        var ttlSeconds = _config.Jwt.RefreshTtlSeconds;
        var token = SignRefreshJwt(userId, jti, ttlSeconds);
        var expiresAt = DateTime.UtcNow.AddSeconds(ttlSeconds);

        _db.RefreshTokens.Add(new RefreshTokenEntity
        {
            Id = jti,
            UserId = userId,
            TokenHash = HashUtil.Sha256Hex(token),
            ExpiresAt = expiresAt,
            CreatedAt = DateTime.UtcNow,
            UserAgent = userAgent is null ? null : userAgent[..Math.Min(userAgent.Length, 255)],
        });
        await _db.SaveChangesAsync(cancellationToken);

        return new IssuedRefreshToken(token, expiresAt);
    }

    public async Task<(Guid UserId, IssuedRefreshToken RefreshToken)> RotateRefreshTokenAsync(
        string rawToken,
        string? userAgent,
        CancellationToken cancellationToken = default)
    {
        var payload = VerifyRefreshJwt(rawToken);
        var tokenHash = HashUtil.Sha256Hex(rawToken);

        var stored = await _db.RefreshTokens
            .AsNoTracking()
            .FirstOrDefaultAsync(t => t.TokenHash == tokenHash, cancellationToken);

        if (stored is null)
        {
            if (Guid.TryParse(payload.Sub, out var missingUserId))
            {
                await RevokeAllForUserAsync(missingUserId, cancellationToken);
            }

            _logger.LogWarning("Refresh token for user {UserId} was not found; all sessions revoked.", payload.Sub);
            throw new UnauthorizedException(ErrorCodes.InvalidRefreshToken, "Refresh token is not valid");
        }

        if (stored.RevokedAt is not null)
        {
            await RevokeAllForUserAsync(stored.UserId, cancellationToken);
            _logger.LogWarning("Reuse of a revoked refresh token by user {UserId}; all sessions revoked.", stored.UserId);
            throw new UnauthorizedException(ErrorCodes.InvalidRefreshToken, "Refresh token has already been used");
        }

        if (stored.ExpiresAt <= DateTime.UtcNow)
        {
            throw new UnauthorizedException(ErrorCodes.InvalidRefreshToken, "Refresh token has expired");
        }

        await using var tx = await _db.Database.BeginTransactionAsync(cancellationToken);
        await _db.RefreshTokens
            .Where(t => t.Id == stored.Id)
            .ExecuteUpdateAsync(s => s.SetProperty(t => t.RevokedAt, DateTime.UtcNow), cancellationToken);

        var refresh = await IssueRefreshTokenAsync(stored.UserId, userAgent, cancellationToken);
        await tx.CommitAsync(cancellationToken);
        return (stored.UserId, refresh);
    }

    public async Task RevokeRefreshTokenAsync(string rawToken, CancellationToken cancellationToken = default)
    {
        var hash = HashUtil.Sha256Hex(rawToken);
        await _db.RefreshTokens
            .Where(t => t.TokenHash == hash && t.RevokedAt == null)
            .ExecuteUpdateAsync(s => s.SetProperty(t => t.RevokedAt, DateTime.UtcNow), cancellationToken);
    }

    public async Task RevokeAllForUserAsync(Guid userId, CancellationToken cancellationToken = default)
    {
        await _db.RefreshTokens
            .Where(t => t.UserId == userId && t.RevokedAt == null)
            .ExecuteUpdateAsync(s => s.SetProperty(t => t.RevokedAt, DateTime.UtcNow), cancellationToken);
    }

    public async Task<int> DeleteExpiredTokensAsync(CancellationToken cancellationToken = default)
    {
        return await _db.RefreshTokens
            .Where(t => t.ExpiresAt < DateTime.UtcNow)
            .ExecuteDeleteAsync(cancellationToken);
    }

    private string SignRefreshJwt(Guid userId, Guid jti, int ttlSeconds)
    {
        var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(_config.Jwt.RefreshSecret));
        var credentials = new SigningCredentials(key, SecurityAlgorithms.HmacSha256);
        var claims = new[]
        {
            new Claim(JwtRegisteredClaimNames.Sub, userId.ToString()),
            new Claim(JwtRegisteredClaimNames.Jti, jti.ToString()),
            new Claim("type", "refresh"),
        };

        var token = new JwtSecurityToken(
            claims: claims,
            expires: DateTime.UtcNow.AddSeconds(ttlSeconds),
            signingCredentials: credentials);

        return new JwtSecurityTokenHandler().WriteToken(token);
    }

    private RefreshTokenPayload VerifyRefreshJwt(string token)
    {
        try
        {
            var handler = new JwtSecurityTokenHandler();
            var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(_config.Jwt.RefreshSecret));
            var principal = handler.ValidateToken(
                token,
                new TokenValidationParameters
                {
                    ValidateIssuer = false,
                    ValidateAudience = false,
                    ValidateIssuerSigningKey = true,
                    IssuerSigningKey = key,
                    ClockSkew = TimeSpan.FromSeconds(30),
                },
                out _);

            if (principal.FindFirst("type")?.Value != "refresh")
            {
                throw new UnauthorizedException(ErrorCodes.InvalidRefreshToken, "Wrong token type");
            }

            var sub = principal.FindFirst(JwtRegisteredClaimNames.Sub)?.Value;
            var jti = principal.FindFirst(JwtRegisteredClaimNames.Jti)?.Value;
            if (sub is null || jti is null || !Guid.TryParse(sub, out var userId))
            {
                throw new UnauthorizedException(ErrorCodes.InvalidRefreshToken, "Refresh token is not valid");
            }

            return new RefreshTokenPayload(sub, jti, "refresh");
        }
        catch (UnauthorizedException)
        {
            throw;
        }
        catch
        {
            throw new UnauthorizedException(ErrorCodes.InvalidRefreshToken, "Refresh token is not valid");
        }
    }
}
